plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -858,29 +870,25 @@ function ( $acc, $query ) {
 				if ( empty( array_intersect( $this->get_valid_query_vars(), array_keys( $query ) ) ) ) {
 					return $this->merge_queries( $acc, ...array_values( $query ) );
 				}
-				return $this->array_merge_recursive_replace_non_array_properties( $acc, $query );
+
+				$result = $this->array_merge_recursive_replace_non_array_properties( $acc, $query );
+
+				// The post__in query needs special handling, as it should be an
+				// intersection of all post__in queries.
+				if (
+					is_array( $acc['post__in'] ) &&
+					! empty( $acc['post__in'] ) &&
+					is_array( $query['post__in'] ) &&
+					! empty( $query['post__in'] )
```

Comment:
I'm getting two warnings on the frontend: `Undefined array key "post__in"` from line 879 and 881, I guess we should check if `$acc['post__in']` and `$query['post__in']` exist first.